File: vision-ai-frontend/src/components/service/cameraService.ts
import { api } from "../../lib/api";

export interface CameraData {
  id?: number;
  name?: string;
  ip_address?: string;
  resolution?: string;
  fps?: number;
  rotation_angle?: number;
  control_mode?: 'auto' | 'manual';
  exposure?: number;
  gain?: number;
  focus?: number;
  brightness?: number;
  contrast?: number;
  auto_focus?: boolean;
  auto_exposure?: boolean;
  is_enabled?: boolean;
}

export const cameraService = {
  async checkCamera() {
    const res = await api.get("/camera/");
    return res.data;
  },

  async createCamera(data: CameraData) {
    const res = await api.post("/camera/create", data);
    return res.data;
  },

  async enableCamera(id: string | number) {
    const res = await api.post(`/camera/enable/${id}`);
    return res.data;
  },

  async disableCamera(id: string | number) {
    const res = await api.post(`/camera/disable/${id}`);
    return res.data;
  },

  async updateCamera(id: string | number, data: CameraData) {
    const res = await api.put(`/camera/update/${id}`, data);
    return res.data;
  },

  async saveCamera(data: CameraData) {
    if (data.id) {
      return await this.updateCamera(data.id, data);
    }
    return await this.createCamera(data);
  },

  async start() {
    const res = await api.post("/oak/start");
    return res.data;
  },

  async stop() {
    const res = await api.post("/oak/stop");
    return res.data;
  },

  async startStream() {
    const res = await api.post("/oak/stream/start");
    return res.data;
  },

  async stopStream() {
    const res = await api.post("/oak/stream/stop");
    return res.data;
  },

  async startLiveInference(sessionId: string = "live") {
    const res = await api.post(`/oak/inference/start/${sessionId}`);
    return res.data;
  },

  async stopLiveInference() {
    const res = await api.post("/oak/inference/stop");
    return res.data;
  },

  async health() {
    const res = await api.get("/oak/health");
    return res.data;
  },

  async updateLiveControls(id: string | number, data: any) {
    const res = await api.patch(`/camera/live-controls/${id}`, data);
    return res.data;
  },

   async updateInferenceMode(mode: "testing" | "production") {
    const res = await api.patch("/camera/inference-mode", { mode });
    return res.data;
  },
};

export default cameraService;

File: vision-ai-backend/app/services/camera_service.py
# app/services/camera_service.py

from sqlalchemy.orm import Session
from fastapi import HTTPException

from app.models.camera_model import Camera


from app.schemas.camera_schema import CameraCreate, CameraUpdate
from app.schemas.basic_config_update_schema import BasicConfigUpdate

from app.core.logger import setup_logger
from app.services.realtime_log_service import realtime_log_service

logger = setup_logger("camera-service")


# =========================================================
# CAMERA CRUD
# =========================================================

def create_camera(db: Session, data: CameraCreate):
    try:
        logger.info("[CREATE_CAMERA] Called")

        camera = Camera(**data.model_dump())

        # Apply control mode rules
        if camera.control_mode == "auto":
            camera.exposure = None
            camera.gain = None
            camera.focus = None
        else:
            camera.auto_exposure = False
            camera.auto_focus = False

        db.add(camera)
        db.commit()
        db.refresh(camera)

        logger.info(f"[CREATE_CAMERA] Created camera ID: {camera.id}")
        realtime_log_service.add_log(
            "camera",
            "CAMERA",
            f"Camera created: {camera.name}",
            "success"
        )
        return camera
    
    except Exception as e:
            logger.error(f"[CREATE_CAMERA] Failed error: {e}", exc_info=True)
            realtime_log_service.add_log(
                "camera",
                "ERROR",
                "Failed to create camera",
                "error"
            )
   


def update_camera_partial(db: Session, camera_id: int, data: CameraUpdate):
    try:
        logger.info(f"[UPDATE_CAMERA] Called with ID: {camera_id}")

        camera = db.query(Camera).filter(Camera.id == camera_id).first()

        if not camera:
            raise HTTPException(status_code=404, detail="Camera not found")

        update_data = data.model_dump(exclude_unset=True)

        for key, value in update_data.items():
            setattr(camera, key, value)

        # Apply control mode AFTER update
        if camera.control_mode == "auto":
            camera.exposure = None
            camera.gain = None
            camera.focus = None
            camera.auto_exposure = True
            camera.auto_focus = True
        elif camera.control_mode == "manual":
            camera.auto_exposure = False
            camera.auto_focus = False

        db.commit()
        db.refresh(camera)

        logger.info("[UPDATE_CAMERA] Success")
        realtime_log_service.add_log(
            "camera",
            "CAMERA",
            f"Camera updated: ID {camera_id}",
            "success"
        )
        return camera

    
    except Exception as e:

        logger.error(f"[UPDATE_CAMERA] Failed error: {e}", exc_info=True)
        realtime_log_service.add_log(
            "camera",
            "WARN",
            f"Camera not found: ID {camera_id}",
            "warning"
        )

def enable_camera(db: Session, camera_id: int):
    try:
        camera = db.query(Camera).filter(Camera.id == camera_id).first()
        if not camera:
            raise HTTPException(status_code=404, detail="Camera not found")

        camera.is_enabled = True
        db.commit()
        db.refresh(camera)
        realtime_log_service.add_log(
            "camera",
            "CAMERA",
            f"Camera enabled: ID {camera_id}",
            "success"
        )
        return camera
    
    except Exception as e:
        logger.error(f"[ENABLE_CAMERA] Failed error: {e}", exc_info=True)
        realtime_log_service.add_log(
            "camera",
            "WARN",
            f"Enable failed - Camera not found: ID {camera_id}",
            "warning"
        )


def disable_camera(db: Session, camera_id: int):
    try:
        camera = db.query(Camera).filter(Camera.id == camera_id).first()
        if not camera:
            raise HTTPException(status_code=404, detail="Camera not found")

        camera.is_enabled = False
        db.commit()
        db.refresh(camera)
        realtime_log_service.add_log(
            "camera",
            "CAMERA",
            f"Camera disabled: ID {camera_id}",
            "success"
        )
        return camera
    
    except Exception as e:
        logger.error(f"[DISABLE_CAMERA] Failed error: {e}", exc_info=True)
        realtime_log_service.add_log(
            "camera",
            "WARN",
            f"Disable failed - Camera not found: ID {camera_id}",
            "warning"
        )


def get_cameras(db: Session):
    return db.query(Camera).all()


def get_camera_config(db: Session):
    return db.query(Camera).first()


def get_inference_config(db: Session):
    return {"mode": "production"}


def update_inference_mode(db: Session, mode: str):
    return {"message": "Inference mode updated", "mode": mode}

def update_basic_config(db: Session, data: BasicConfigUpdate):
    logger.info("[UPDATE_BASIC_CONFIG] Started")
    realtime_log_service.add_log(
        "system",
        "CONFIG",
        "Updating system configuration",
        "info"
    )

    try:
        # =========================================================
        # CAMERA
        # =========================================================
        camera = db.query(Camera).first()

        if not camera:
            logger.info("[CONFIG] Creating Camera")

            if not data.camera or not data.camera.name:
                raise HTTPException(status_code=400, detail="Camera name is required")

            camera_data = data.camera.model_dump(exclude_unset=True)
            camera = Camera(**camera_data)
            db.add(camera)

        else:
            if data.camera:
                logger.info("[CONFIG] Updating Camera")

                cam_data = data.camera.model_dump(exclude_unset=True)
                for key, value in cam_data.items():
                    setattr(camera, key, value)

        # Apply control mode AFTER update/create
        if camera.control_mode == "auto":
            camera.exposure = None
            camera.gain = None
            camera.focus = None
            camera.auto_exposure = True
            camera.auto_focus = True
        elif camera.control_mode == "manual":
            camera.auto_exposure = False
            camera.auto_focus = False

        # =========================================================
        # TRAINING
        # =========================================================
        

        # =========================================================
        # INFERENCE
        # =========================================================
        

        db.commit()

        logger.info("[UPDATE_BASIC_CONFIG] Success")
        realtime_log_service.add_log(
            "system",
            "CONFIG",
            "Configuration updated successfully",
            "success"
        )

        return {
            "message": "Config updated successfully"
        }

    except Exception as e:
        db.rollback()
        logger.error(f"[UPDATE_BASIC_CONFIG ERROR] {e}", exc_info=True)
        realtime_log_service.add_log(
            "system",
            "ERROR",
            "Configuration update failed",
            "error"
        )
        raise HTTPException(status_code=500, detail=str(e))


# =========================================================
# GET CONFIG
# =========================================================

def get_basic_config(db: Session):
    logger.info("[GET_BASIC_CONFIG] Fetching configs")
    realtime_log_service.add_log(
        "system",
        "CONFIG",
        "Fetching configuration",
        "info"
    )

    camera = db.query(Camera).first()
    

    return {
        "camera": camera,
        
        "inference": {"mode": "production"}
    }